diff --git a/stage1/00-boot-files/00-run.sh b/stage1/00-boot-files/00-run.sh
index bc61397..d71e260 100755
--- a/stage1/00-boot-files/00-run.sh
+++ b/stage1/00-boot-files/00-run.sh
@@ -2,3 +2,4 @@
 
 install -m 644 files/cmdline.txt "${ROOTFS_DIR}/boot/"
 install -m 644 files/config.txt "${ROOTFS_DIR}/boot/"
+install -m 644 files/firstrun.sh "${ROOTFS_DIR}/boot/"
diff --git a/stage1/00-boot-files/files/cmdline.txt b/stage1/00-boot-files/files/cmdline.txt
index 03cde68..cd6cc93 100644
--- a/stage1/00-boot-files/files/cmdline.txt
+++ b/stage1/00-boot-files/files/cmdline.txt
@@ -1 +1 @@
-console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait
+console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait systemd.run=/boot/firstrun.sh
diff --git a/stage1/00-boot-files/files/firstrun.sh b/stage1/00-boot-files/files/firstrun.sh
new file mode 100755
index 0000000..bc91234
--- /dev/null
+++ b/stage1/00-boot-files/files/firstrun.sh
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set +e
+
+if [ -f "/boot/tower.env" ]; then
+    . "/boot/tower.env"
+
+    /usr/lib/raspberrypi-sys-mods/imager_custom set_hostname "$NAME.tower"
+    /usr/lib/raspberrypi-sys-mods/imager_custom enable_ssh -k "$PUBLIC_KEY"
+    /usr/lib/userconf-pi/userconf "$USER" "$ENCRYPTED_PASSWORD"
+    /usr/lib/raspberrypi-sys-mods/imager_custom set_keymap "$KEYMAP"
+    /usr/lib/raspberrypi-sys-mods/imager_custom set_timezone "$TIMEZONE"
+
+    if "$ONLINE" == "true"; then
+        /usr/lib/raspberrypi-sys-mods/imager_custom set_wlan "$WLAN_SSID" "$WLAN_PASSWORD" "$WLAN_COUNTRY"
+    fi
+
+    rm -f /boot/tower.env
+fi
+
+rm -f /boot/firstrun.sh
+sed -i 's| systemd.run.*||g' /boot/cmdline.txt
+exit 0
diff --git a/stage2/01-sys-tweaks/00-patches/07-resize-init.diff b/stage2/01-sys-tweaks/00-patches/07-resize-init.diff
index dfc01d4..9ea222c 100644
--- a/stage2/01-sys-tweaks/00-patches/07-resize-init.diff
+++ b/stage2/01-sys-tweaks/00-patches/07-resize-init.diff
@@ -1,5 +1,5 @@
 --- stage2.orig/rootfs/boot/cmdline.txt
 +++ stage2/rootfs/boot/cmdline.txt
 @@ -1 +1 @@
--console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait
-+console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait quiet init=/usr/lib/raspberrypi-sys-mods/firstboot
+-console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait systemd.run=/boot/firstrun.sh
++console=serial0,115200 console=tty1 root=ROOTDEV rootfstype=ext4 fsck.repair=yes rootwait quiet systemd.run=/boot/firstrun.sh init=/usr/lib/raspberrypi-sys-mods/firstboot
diff --git a/stagetower/00-install-x2goserver/00-packages b/stagetower/00-install-x2goserver/00-packages
new file mode 100644
index 0000000..22223ce
--- /dev/null
+++ b/stagetower/00-install-x2goserver/00-packages
@@ -0,0 +1,2 @@
+x2goserver
+x2goserver-xsession
diff --git a/stagetower/01-install-apt-offline/00-packages b/stagetower/01-install-apt-offline/00-packages
new file mode 100644
index 0000000..f40b64c
--- /dev/null
+++ b/stagetower/01-install-apt-offline/00-packages
@@ -0,0 +1 @@
+apt-offline
\ No newline at end of file
diff --git a/stagetower/EXPORT_IMAGE b/stagetower/EXPORT_IMAGE
new file mode 100644
index 0000000..b5e2cc2
--- /dev/null
+++ b/stagetower/EXPORT_IMAGE
@@ -0,0 +1,4 @@
+IMG_SUFFIX="-tower"
+if [ "${USE_QEMU}" = "1" ]; then
+	export IMG_SUFFIX="${IMG_SUFFIX}-qemu"
+fi
diff --git a/stagetower/EXPORT_NOOBS b/stagetower/EXPORT_NOOBS
new file mode 100644
index 0000000..ecd98dc
--- /dev/null
+++ b/stagetower/EXPORT_NOOBS
@@ -0,0 +1,2 @@
+NOOBS_NAME="Raspberry Pi OS for Tower system (64-bit)"
+NOOBS_DESCRIPTION="A port of Debian with desktop and recommended applications"
diff --git a/stagetower/prerun.sh b/stagetower/prerun.sh
new file mode 100755
index 0000000..9acd13c
--- /dev/null
+++ b/stagetower/prerun.sh
@@ -0,0 +1,5 @@
+#!/bin/bash -e
+
+if [ ! -d "${ROOTFS_DIR}" ]; then
+	copy_previous
+fi
